core/src/main/java/org/apache/calcite/rel/rules/ReduceExpressionsRule.java

```diff
@@ -302,6 +302,7 @@ public ProjectReduceExpressionsRule(Class<? extends Project> projectClass,
           Lists.newArrayList(project.getProjects());
       if (reduceExpressions(project, expList, predicates, false,
           matchNullability)) {
+        assert !project.getProjects().equals(expList);
```

Comment:
Can you please add a message, so the failure looks human-readable?

Reply:
Ok. Will do.